src/skelet_cps.rs

```diff
@@ -0,0 +1,376 @@
+// reimplementation of reimplementation of skelet's implementation :) https://gist.github.com/savask/1c43a0e5cdd81229f236dcf2b0611c3f
+
+use anyhow::{Context, Result};
+use hashbrown::HashMap;
+use indexmap::IndexSet;
+use std::{fmt, str};
+
+use crate::{
+    interner::{ITape, InternedDisplay, InternerTape},
+    machine::{Head, Machine, Orientation},
+    ui_dbg, ProverResult,
+};
+
+#[derive(Debug)]
+pub enum Err {
+    Halt,
+    Loop,
+    ConfLimit,
+}
+
+type Interner = InternerTape<u8>;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+struct Configuration {
+    /// 0 == left, 1 == right, 2 == middle
+    tape: [ITape; 3],
+    head: Head,
+}
+
+impl Configuration {
+    pub fn new(zeros: ITape) -> Configuration {
+        // 0..0 0..0 A> 0..0
+        Configuration { tape: [zeros, zeros, zeros], head: Head { state: 0, orient: 1 } }
+    }
+
+    /// runs until head leaves the tape, split & return new conf, C (`<S ABC` -> `<S AB` & `C`) & end conf string if in tui mode
+    fn run(
+        mut self,
+        machine: &Machine,
+        mut step_limit: usize,
+        segment_size: usize,
+        interner: &mut Interner,
+    ) -> Result<(Configuration, ITape, String), Err> {
+        // TODO(perf): switch to single tape?
+        let mut tape = [interner[self.tape[0]].to_vec(), interner[self.tape[1]].to_vec()];
+        tape[self.head.op_orient()].extend_from_slice(&interner[self.tape[2]]);
+        fn format_conf(tape: &[Vec<u8>; 2], head: Head) -> String {
+            format!(
+                "{}{}{}",
+                tape[0].iter().map(u8::to_string).collect::<String>(),
+                head,
+                tape[1].iter().rev().map(u8::to_string).collect::<String>(),
+            )
+        }
+        while let Some(symbol) = tape[self.head.orient as usize].pop() {
+            let trans = machine.get_transition(symbol, self.head.state).ok_or(Err::Halt)?;
+            if trans.head.state >= machine.states() {
+                return Err(Err::Halt);
+            }
+            self.head = trans.head;
+            tape[1 - self.head.orient as usize].push(trans.symbol);
```

Comment:
good catch, i even [implemented it](https://github.com/univerz/bbc/commit/0122fc59b4991c76f69d4c2cadb077f331ad50bb#diff-dd97281cbaa3b13ed5a8dc4417797959632eed864b774977e013fc307dcc9f01R30) for this prover, because it uses it a lot, but forget to change this one.

Reply:
> `other_dir`

is `direction` better name than `orientation` for head.foo in our "two stacks" tape implementation?

Reply:
> is `direction` better name than `orientation` for head.foo in our "two stacks" tape implementation?

I like `direction` better and I think it's used more commonly. But `orientation` is also a reasonable name. I guess if I think more deeply about it, movement would always use the work "direction" (the TM moved to the Left and again in the same direction). Orientation makes sense in English for a static situation (the TM faced Left and was later in the same orientation). But I wouldn't use "orientation" for movement I don't think.

Poking around a few other codebases I see that Nathan (https://github.com/Nathan-Fenner/bbchallenge-dafny-deciders/blob/main/cyclers-decider.dfy#L62) and savask (I think: https://gist.github.com/savask/1c43a0e5cdd81229f236dcf2b0611c3f#file-turing-hs-L135) also use "direction".